SIL.Core/ClearShare/OlacSystem.cs

```diff
@@ -96,23 +96,52 @@ public string GetXmlForWork(Work work)
 		/// Get all the roles in the system's controlled vocabulary
 		/// </summary>
 		/// ------------------------------------------------------------------------------------
-		public IEnumerable<Role> GetRoles()
+		public IEnumerable<Role> GetRoles(string customRolesXmlFilename = null)
 		{
+			List<Role> ReadRoles(Stream stream)
+			{
+				List<Role> roles;
+				using (var xmlReader = XmlReader.Create(stream))
+				{
+					var doc = XDocument.Load(xmlReader);
+
+					// This is a bit confusing because the role heading node is at the same level
+					// (i.e. a sibling of) as all the associated term nodes. Therefore, the first
+					// thing to do is get the heading nodes that are children of section nodes.
+					// Then find the one heading node whose content is "Role". Then backup to the
+					// section containing that heading node and take all "term" child nodes of
+					// that section (i.e. that are siblings of the role heading node).
+					roles = doc.Descendants("body").Descendants("section")
+						.Elements("heading")
+						.Where(n => n.Value == "Role").Ancestors("section").First()
+						.Descendants("term")
+						.Select(n => new Role(n.Element("code").Value,
+							n.Element("name").Value, n.Element("definition").Value)).ToList();
+					xmlReader.Close();
+				}
+
+				stream.Close();
+				stream.Dispose();
+
+				return roles;
+			}
+
 			if (_roles == null)
 			{
-				// TODO: Provide a way for user-specified roles to be read from a roles.xml
-				// file in a folder somewhere related to the application.
-				var doc = XDocument.Parse(Resources.OlacRoles);
-
-				// This is a bit confusing because the role heading node is at the same level
-				// (i.e. a sibling of) as all the associated term nodes. Therefore, the first
-				// thing to do is get the heading nodes that are children of section nodes.
-				// Then find the one heading node whose content is "Role". Then backup to the
-				// section containing that heading node and take all "term" child nodes of
-				// that section (i.e. that are siblings of the role heading node).
-				_roles = doc.Descendants("body").Descendants("section").Elements("heading")
-						.Where(n => n.Value == "Role").Ancestors("section").First().Descendants("term")
-						.Select(n => new Role(n.Element("code").Value, n.Element("name").Value, n.Element("definition").Value));
+				if (customRolesXmlFilename != null)
+				{
+					try
+					{
+						_roles = ReadRoles(new FileStream(customRolesXmlFilename, FileMode.Open));
+					}
+					catch (Exception e)
+					{
+						Console.WriteLine(e);
```

Comment:
Yeah, I went back and forth on this. I was implementing a TODO, so currently no one is relying on ths, so it's probably YAGNI. If the caller-supplied list is coming from a developer, then they should still be able to catch the exception or otherwise note that the results do not contain the roles they were hoping for. If it comes from a user, then if they pass in something invalid, I'm not sure we'd want the program to crash. I've changed it to use the error reporting and report it as non-fatal.